File: apps-microservices/graph-rag-api-recherche-rust-service/src/routers/nodes.rs
use actix_web::{get, put, web, HttpResponse};
use serde_json::json;

use crate::domain::models::NodeUpdateRequest;
use crate::services::node_service::NodeService;

/// PUT /nodes/{label}/{id} — Update a node's properties.
#[utoipa::path(
    put,
    path = "/nodes/{label}/{id}",
    params(
        ("label" = String, Path, description = "Node label"),
        ("id" = String, Path, description = "Node ID")
    ),
    request_body = NodeUpdateRequest,
    responses(
        (status = 200, description = "Updated node")
    ),
    tag = "Admin"
)]
#[put("/nodes/{label}/{id}")]
pub async fn update_node(
    path: web::Path<(String, String)>,
    body: web::Json<NodeUpdateRequest>,
) -> HttpResponse {
    let (label, id) = path.into_inner();
    let (success, data, _error) = NodeService::update_node(&label, &id, &body.properties).await;
    if success {
        HttpResponse::Ok().json(json!({
            "message": format!("Node {} {} updated successfully.", label, id),
            "node": data,
        }))
    } else {
        HttpResponse::NotFound().json(json!({
            "detail": format!("Node {} with ID {} not found.", label, id)
        }))
    }
}

/// GET /nodes/{label} — Get schema for a node label.
#[utoipa::path(
    get,
    path = "/nodes/{label}",
    params(("label" = String, Path, description = "Node label")),
    responses(
        (status = 200, description = "Node schema")
    ),
    tag = "Admin"
)]
#[get("/nodes/{label}")]
pub async fn get_schema(path: web::Path<String>) -> HttpResponse {
    let label = path.into_inner();
    let schema = NodeService::get_schema(&label).await;
    HttpResponse::Ok().json(json!({
        "label": label,
        "schema": schema,
    }))
}

/// GET /nodes/{label}/{id} — Get a specific node.
#[utoipa::path(
    get,
    path = "/nodes/{label}/{id}",
    params(
        ("label" = String, Path, description = "Node label"),
        ("id" = String, Path, description = "Node ID")
    ),
    responses(
        (status = 200, description = "Node data")
    ),
    tag = "Admin"
)]
#[get("/nodes/{label}/{id}")]
pub async fn get_node(path: web::Path<(String, String)>) -> HttpResponse {
    let (label, id) = path.into_inner();
    match NodeService::get_node(&label, &id).await {
        Some(node) => HttpResponse::Ok().json(json!({
            "code": 200,
            "data": {
                "label": label,
                "id": id,
                "node": node,
            },
        })),
        None => HttpResponse::Ok().json(json!({
            "code": 404,
            "data": {
                "label": label,
                "id": id,
                "node": null,
            },
        })),
    }
}

File: apps-microservices/graph-rag-api-admin-service/app/routers/nodes.py
from fastapi import APIRouter, HTTPException, status
from typing import Dict, Any, Optional
import logging

from app.domain.models import (
    BatchGetRequest,
    BatchResponse,
    BatchUpdateRequest,
    BatchUpsertRequest,
)
from app.services.node_service import node_service

router = APIRouter()


@router.post("/{label}/batch/get", response_model=BatchResponse)
async def batch_get_nodes_route(label: str, body: BatchGetRequest):
    """
    Batch get nodes by label.

    Executes a SINGLE Cypher query (`MATCH ... WHERE n.id IN $ids`) — not N
    queries — so the cost is one index scan, regardless of batch size.

    - **label**: Node label (e.g., 'Produit', 'Fournisseur', 'Categorie')
    - **body.ids**: Raw IDs (without label prefix), 1..500 items.

    Returns `{"found": [...], "missing": [...]}`. IDs not present in the
    database appear in `missing`; they do NOT cause the request to fail.
    """
    try:
        return await node_service.batch_get_nodes(label, body.ids)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(
            f"Error in POST /nodes/{label}/batch/get: {e}", exc_info=True
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An internal error occurred during batch GET.",
        )


@router.post("/{label}/batch/upsert", response_model=BatchResponse)
async def batch_upsert_nodes_route(label: str, body: BatchUpsertRequest):
    """
    Batch UPSERT: apply the SAME properties to many existing nodes.

    Executes a SINGLE Cypher query (`MATCH ... WHERE n.id IN $ids SET n += $props`)
    in a single transaction. Match-only (no node creation): IDs that do not
    exist are returned in `missing`.

    - **label**: Node label (e.g., 'Produit', 'Fournisseur', 'Categorie')
    - **body.ids**: Raw IDs (without label prefix), 1..500 items.
    - **body.properties**: Properties merged into every matched node.

    Use this instead of `/batch/update` when every node should receive the
    same patch (e.g., bulk status change).
    """
    try:
        return await node_service.batch_upsert_nodes(
            label, body.ids, body.properties
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(
            f"Error in POST /nodes/{label}/batch/upsert: {e}", exc_info=True
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An internal error occurred during batch UPSERT.",
        )


@router.post("/{label}/batch/update", response_model=BatchResponse)
async def batch_update_nodes_route(label: str, body: BatchUpdateRequest):
    """
    Batch update nodes by label.

    Executes a SINGLE Cypher query (`UNWIND ... MATCH ... SET n += props`) in
    a single transaction. Any Cypher-level failure rolls back the entire
    batch (all-or-nothing).

    - **label**: Node label (e.g., 'Produit', 'Fournisseur', 'Categorie')
    - **body.items**: List of `{id, properties}` items, 1..500 entries.

    Returns `{"found": [...], "missing": [...]}`. IDs not matched are listed
    in `missing`; matched nodes have their properties merged (`SET n +=`).
    """
    try:
        return await node_service.batch_update_nodes(label, body.items)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(
            f"Error in POST /nodes/{label}/batch/update: {e}", exc_info=True
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An internal error occurred during batch UPDATE.",
        )


@router.put("/{label}/{id}", response_model=Dict[str, Any])
async def update_node(label: str, id: str, properties: Dict[str, Any]):
    """
    Update a specific node by its label and ID.

    - **label**: The node label (e.g., 'Produit', 'Fournisseur')
    - **id**: The unique identifier of the node.
    - **properties**: A dictionary of properties to update.
    """
    try:
        updated_node = await node_service.update_node(label, id, properties)

        if updated_node:
            return {
                "message": f"Node {label} {id} updated successfully.",
                "node": updated_node,
            }
        else:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Node {label} with ID {id} not found.",
            )

    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(f"Error in PUT /nodes/{label}/{id}: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An internal error occurred while updating the node.",
        )


@router.get("/{label}", response_model=Dict[str, Any])
async def get_node_schema(label: str):
    """
    Get the schema (properties and datatypes) for a specific node label.

    - **label**: The node label (e.g., 'Produit', 'Fournisseur')
    """
    try:
        schema = await node_service.get_node_schema(label)

        if schema:
            return {
                "label": label,
                "schema": schema,
            }
        else:
            # If empty, it might mean label doesn't exist or has no properties.
            # We return empty schema but 200 OK is usually better than 404 if the label is just not found in schema cache yet,
            # but usually getting schema for non-existent label returns empty.
            # Let's return what we found.
            return {
                "label": label,
                "schema": [],
                "message": f"No schema found for label {label}. It might not exist in the database yet.",
            }

    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(f"Error in GET /nodes/{label}: {e}", exc_info=True)
        raise HTTPException(
            detail="An internal error occurred while fetching the node schema.",
        )


@router.get("/{label}/{id}", response_model=Dict[str, Any])
async def get_node(label: str, id: str):
    """
    Get a specific node by its label and ID.

    - **label**: The node label (e.g., 'Produit', 'Fournisseur')
    - **id**: The unique identifier of the node.
    """
    try:
        node = await node_service.get_node(label, id)

        if node:
            return {
                "code": 200,
                "data": {
                    "label": label,
                    "id": id,
                    "node": node,
                },
            }
        else:
            return {
                "code": 404,
                "data": {
                    "label": label,
                    "id": id,
                    "node": None,
                },
            }

    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logging.error(f"Error in GET /nodes/{label}/{id}: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An internal error occurred while fetching the node.",
        )
